gnss_lib_py/utils/sv_models.py

```diff
@@ -0,0 +1,611 @@
+    # be used but compared against SP3 and Android data
+
+    ############################################
+    ######  Lines added for velocity (4)  ######
+    ############################################
+    omega_dot = ephem['OmegaDot'] - consts.OMEGA_E_DOT
+    sv_posvel['vx_sv_mps'] = (dxp * cos_omega
+                         - dyp * cos_i*sin_omega
+                         + y_plane  * sin_omega*sin_i*delta_i
+                         - (x_plane * sin_omega + y_plane*cos_i*cos_omega)*omega_dot)
+
+    sv_posvel['vy_sv_mps'] = (dxp * sin_omega
+                         + dyp * cos_i * cos_omega
+                         - y_plane  * sin_i * cos_omega * delta_i
+                         + (x_plane * cos_omega - (y_plane*cos_i*sin_omega)) * omega_dot)
+
+    sv_posvel['vz_sv_mps'] = dyp*sin_i + y_plane*cos_i*delta_i
+
+    return sv_posvel
+
+
+def _filter_ephemeris_measurements(measurements, constellations, ephemeris_path):
+    """Filter measurements based on constellations and ephemeris on received SVs
+
+    Measurements are filtered to contain the intersection of received and
+    desired constellations.
+    Ephemeris is extracted from the given path and filtered to contain
+    received SVs.
+
+    Parameters
+    ----------
+    measurements : gnss_lib_py.parsers.navdata.NavData
+        Recevied measurements that are filtered based on constellations.
+    constellations : list
+        List of strings indicating constellations that we want to use.
+    ephemeris_path : string
+        Path where the ephermis files are stored or downloaded to.
+
+    Returns
+    -------
+    measurements_subset : gnss_lib_py.parsers.navdata.NavData
+        Measurements containing desired constellations
+    ephem : gnss_lib_py.parsers.navdata.NavData
+        Ephemeris parameters for received SVs and constellations
+    """
+    measurements.in_rows(['gnss_id', 'sv_id', 'gps_millis'])
+    # Check whether the rows are in the right format as needed.
+    isinstance(measurements['gnss_id'].dtype, object)
+    isinstance(measurements['sv_id'].dtype, int)
+    isinstance(measurements['gps_millis'].dtype, np.int64)
+    rx_const= np.unique(measurements['gnss_id'])
+    # Check if required constellations are available, keep only required
+    # constellations
+    for const in constellations:
+        if const not in rx_const:
+            warnings.warn(const + " not available in received constellations", RuntimeWarning)
+    rx_const_set = set(rx_const)
+    req_const_set = set(constellations)
+    keep_consts = req_const_set.intersection(rx_const_set)
+
+    measurements_subset = measurements.keep_cols_where('gnss_id', keep_consts, condition="eq")
+
+    # preprocessing of received quantities for downloading ephemeris file
+    eph_sv = _combine_gnss_sv_ids(measurements)
+    lookup_sats = list(np.unique(eph_sv))
+    start_gps_millis = np.min(measurements['gps_millis'])
```

Comment:
Should this be end time instead, since get_ephemeris returns data before the given timestamp?

Reply:
Ah, I see what might have caused the confusion. This is called `start_gps_millis` because it's the start of the trajectory and doesn't really have as much to do with the `EphemerisManager`